x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/common/H3SphericalUtil.java

```diff
@@ -0,0 +1,129 @@
+            // specialize south pole
+            computeSouthPoleBounds(boundary, boundingBox);
+        } else {
+            // generic case
+            computeBounds(boundary, boundingBox);
+        }
+    }
+
+    private static void computeNorthPoleBounds(CellBoundary boundary, GeoBoundingBox boundingBox) {
+        double minLat = Double.POSITIVE_INFINITY;
+        for (int i = 0; i < boundary.numPoints(); i++) {
+            minLat = Math.min(minLat, boundary.getLatLon(i).getLatRad());
+        }
+        boundingBox.topLeft().reset(90, -180d);
+        boundingBox.bottomRight().reset(Math.toDegrees(minLat), 180d);
+    }
+
+    private static void computeSouthPoleBounds(CellBoundary boundary, GeoBoundingBox boundingBox) {
+        double maxLat = Double.NEGATIVE_INFINITY;
+        for (int i = 0; i < boundary.numPoints(); i++) {
+            maxLat = Math.max(maxLat, boundary.getLatLon(i).getLatRad());
+        }
+        boundingBox.topLeft().reset(Math.toDegrees(maxLat), -180d);
+        boundingBox.bottomRight().reset(-90, 180d);
+    }
+
+    private static void computeBounds(CellBoundary boundary, GeoBoundingBox boundingBox) {
+        // This algorithm is based on the bounding box for great circle edges in
+        // https://trs.jpl.nasa.gov/bitstream/handle/2014/41271/07-0286.pdf
+        double minLat = Double.POSITIVE_INFINITY;
+        int minLatPos = -1;
+        double minLon = Double.POSITIVE_INFINITY;
+        double maxLat = Double.NEGATIVE_INFINITY;
+        int maxLatPos = -1;
+        double maxLon = Double.NEGATIVE_INFINITY;
+        double maxNegLon = Double.NEGATIVE_INFINITY;
+        double minPosLon = Double.POSITIVE_INFINITY;
+        for (int i = 0; i < boundary.numPoints(); i++) {
+            final double lon = boundary.getLatLon(i).getLonRad();
+            final double lat = boundary.getLatLon(i).getLatRad();
+            if (lat < minLat) {
+                minLat = lat;
+                minLatPos = i;
+            }
+            if (lat > maxLat) {
+                maxLat = lat;
+                maxLatPos = i;
+            }
+            minLon = Math.min(minLon, lon);
+            maxLon = Math.max(maxLon, lon);
+            if (lon < 0) {
+                maxNegLon = Math.max(maxNegLon, lon);
+            } else {
+                minPosLon = Math.min(minPosLon, lon);
+            }
+        }
+        if (minLat < 0) {
+            // we only correct the min latitude if negative
+            minLat = boundary.getLatLon(minLatPos).greatCircleMinLatitude(computeEdge(boundary, minLatPos, MIN_COMPARATOR));
+        }
+        if (maxLat > 0) {
+            // we only correct the max latitude if positive
+            maxLat = boundary.getLatLon(maxLatPos).greatCircleMaxLatitude(computeEdge(boundary, maxLatPos, MAX_COMPARATOR));
+        }
+        // the min / max longitude is computed the same way as in cartesian, being careful with polygons crossing the dateline
```

Comment:
I guess, as with cartesian, we do not allow geometries wider that 180, except for the polar cells which are 360 degrees. I presume we've verified that only polar cells are wider than 180?

Reply:
Yes, that's correct.